File: corpus/javascript/session-broker/src/envelope.ts
/**
 * Session payload sealing.
 *
 * Broker sessions carry short-lived credentials, so every payload is sealed
 * with a fresh nonce under the session's data key.
 */

import { createCipheriv, createDecipheriv, randomBytes } from 'node:crypto';

export const NONCE_BYTES = 12;
export const DATA_KEY_BYTES = 32;
export const TAG_BYTES = 16;

export interface Sealed {
  nonce: Buffer;
  ciphertext: Buffer;
  tag: Buffer;
}

/** Return a fresh data key for one session. */
export function newDataKey(): Buffer {
  return randomBytes(DATA_KEY_BYTES);
}

/** Seal a session payload. */
export function seal(dataKey: Buffer, plaintext: Buffer, associated: Buffer): Sealed {
  if (dataKey.length !== DATA_KEY_BYTES) {
    throw new Error(`envelope: data key must be ${DATA_KEY_BYTES} bytes`);
  }
  const nonce = randomBytes(NONCE_BYTES);
  const cipher = createCipheriv('aes-256-gcm', dataKey, nonce);
  cipher.setAAD(associated);
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  return { nonce, ciphertext, tag: cipher.getAuthTag() };
}

/** Reverse {@link seal}. */
export function open(dataKey: Buffer, sealed: Sealed, associated: Buffer): Buffer {
  const decipher = createDecipheriv('aes-256-gcm', dataKey, sealed.nonce);
  decipher.setAAD(associated);
  decipher.setAuthTag(sealed.tag);
  return Buffer.concat([decipher.update(sealed.ciphertext), decipher.final()]);
}

/** Pack a sealed payload for the wire. */
export function toWire(sealed: Sealed): Buffer {
  return Buffer.concat([sealed.nonce, sealed.tag, sealed.ciphertext]);
}

/** Unpack a wire payload. */
export function fromWire(blob: Buffer): Sealed {
  if (blob.length <= NONCE_BYTES + TAG_BYTES) {
    throw new Error('envelope: payload too short');
  }
  return {
    nonce: blob.subarray(0, NONCE_BYTES),
    tag: blob.subarray(NONCE_BYTES, NONCE_BYTES + TAG_BYTES),
    ciphertext: blob.subarray(NONCE_BYTES + TAG_BYTES),
  };
}

File: corpus-src/rust/sealbox/src/envelope.rs
//! Blob sealing.
//!
//! Every blob that reaches the object store is sealed first, so a leaked
//! bucket listing is not a leaked blob.

use aes_gcm::aead::{Aead, AeadCore, KeyInit, OsRng};
use aes_gcm::{Aes256Gcm, Key, Nonce};

pub const NONCE_BYTES: usize = 12;
pub const DATA_KEY_BYTES: usize = 32;

/// A sealed blob and the nonce needed to open it.
#[derive(Clone, Debug)]
pub struct Sealed {
    pub nonce: Vec<u8>,
    pub ciphertext: Vec<u8>,
}

impl Sealed {
    /// Pack a sealed blob for the wire.
    pub fn to_wire(&self) -> Vec<u8> {
        let mut out = Vec::with_capacity(self.nonce.len() + self.ciphertext.len());
        out.extend_from_slice(&self.nonce);
        out.extend_from_slice(&self.ciphertext);
        out
    }

    /// Unpack a wire blob.
    pub fn from_wire(blob: &[u8]) -> Result<Self, Error> {
        if blob.len() <= NONCE_BYTES {
            return Err(Error::Malformed);
        }
        Ok(Self {
            nonce: blob[..NONCE_BYTES].to_vec(),
            ciphertext: blob[NONCE_BYTES..].to_vec(),
        })
    }
}

/// What can go wrong while sealing.
#[derive(Debug)]
pub enum Error {
    KeyLength,
    Malformed,
    Aead,
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Error::KeyLength => write!(f, "data key must be {DATA_KEY_BYTES} bytes"),
            Error::Malformed => write!(f, "sealed blob is malformed"),
            Error::Aead => write!(f, "authenticated encryption failed"),
        }
    }
}

impl std::error::Error for Error {}

/// Return a fresh data key for one blob.
pub fn new_data_key() -> Vec<u8> {
    Aes256Gcm::generate_key(&mut OsRng).to_vec()
}

/// Seal a blob.
pub fn seal(data_key: &[u8], plaintext: &[u8]) -> Result<Sealed, Error> {
    if data_key.len() != DATA_KEY_BYTES {
        return Err(Error::KeyLength);
    }
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(data_key)); //@PS rust-l1-aesgcm|AES-256-GCM|1|algorithm|blob AEAD
    let nonce = Aes256Gcm::generate_nonce(&mut OsRng);
    let ciphertext = cipher.encrypt(&nonce, plaintext).map_err(|_| Error::Aead)?;
    Ok(Sealed {
        nonce: nonce.to_vec(),
        ciphertext,
    })
}

/// Reverse [`seal`].
pub fn open(data_key: &[u8], sealed: &Sealed) -> Result<Vec<u8>, Error> {
    if data_key.len() != DATA_KEY_BYTES {
        return Err(Error::KeyLength);
    }
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(data_key));
    let nonce = Nonce::from_slice(&sealed.nonce);
    cipher
        .decrypt(nonce, sealed.ciphertext.as_ref())
        .map_err(|_| Error::Aead)
}
